fix(wallet)!: Improve test utilities

ValuedMammal opened the pull request:
Follow up to #1643, refactor `insert_anchor_from_conf` to just insert an anchor of type `ConfirmationBlockTime`

### Notes to the reviewers

The PR introduces a public `test_utils` module and "test-utils" cargo feature that exposes common helpers such as `get_funded_wallet`. Credit to #1492 for inspiring that idea. Usage of test utilities is enhanced overall, and tests are less dependent on problematic APIs that may be removed in the future.

### Changelog notice

- `bdk_wallet`: Added "test-utils" feature flag that exposes common helpers for testing and development
- Removed methods `Wallet::insert_tx`, `Wallet::insert_checkpoint`, `Wallet::unbroadcast_transactions`

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing
* [x] I've added docs for the new feature
* [x] This pull request breaks the existing API

evanlinjin committed:
revert(wallet)!: rm `Wallet::unbroadcast_transactions`

This is no longer relevant as we direct callers to only insert tx into
the wallet after a successful broadcast.

evanlinjin committed:
feat(wallet)!: make `seen_at` mandatory for `Wallet::apply_update_at`

Not including a `seen_at` alongside the update means the unconfirmed txs
of the update will not be considered to be part of the canonical
history. Therefore, transactions created with this wallet may replace
the update's unconfirmed txs (which is unintuitive behavior).

Also updated the docs.

ValuedMammal committed:
test(wallet): refactor helper `insert_anchor_from_conf`

ValuedMammal commented on crates/wallet/tests/common.rs:
```diff
@@ -208,23 +208,17 @@ pub fn feerate_unchecked(sat_vb: f64) -> FeeRate {
-                    anchors: [(anchor, txid)].into(),
-                    ..Default::default()
-                },
+/// Simulates confirming a tx with `txid` by applying an update to the wallet containing
+/// the given `anchor`. Note: to be considered confirmed the anchor block must exist in
+/// the current active chain.
+pub fn insert_anchor(wallet: &mut Wallet, txid: Txid, anchor: ConfirmationBlockTime) {
+    wallet
```
Can possibly add an assertion here that the anchor height is not greater than the current chain tip.

oleonardolima commented on crates/wallet/tests/wallet.rs:
```diff
@@ -4113,7 +4109,7 @@ fn test_keychains_with_overlapping_spks() {
     let chain_position = ChainPosition::Confirmed(ConfirmationBlockTime {
         block_id: BlockId {
-            height: 8000,
+            height: 2000,
```
I might've not noticed it properly, but why was this change necessary?

ValuedMammal replied:
I think the height of the anchor block was inadvertently changed in #1643 and this changes it back to the way it was as of 1.0.0-beta.5

evanlinjin replied:
@ValuedMammal Woops, thanks for picking up on that!

oleonardolima reviewed (commented):
Concept ACK
Although I didn't get it yet why the change, mentioned in the comment below.

ValuedMammal committed:
feat(wallet): add module `test_utils`

The common test utils are moved to a public `test_utils` module
behind a new test-utils feature flag

ValuedMammal committed:
test(wallet): add helpers to `test_utils`

ValuedMammal committed:
test(wallet): fix test descriptor getters

- `get_funded_wallet` requires two descriptors
- `get_funded_wallet_single` returns a single-descriptor
wallet

ValuedMammal committed:
chore: fix imports

ValuedMammal committed:
test(wallet): improve usage of test utils

notmandatory commented:
Is this an API break change or something that should go in the 1.0-beta milestone?

ValuedMammal committed:
fix(wallet)!: delete method `insert_tx`

Inserting unconfirmed txs can be done using the existing method
`apply_unconfirmed_txs`. Also removed `insert_checkpoint`, as the
API is unclear regarding where in the local chain the given block
should connect. Analogs for these methods are found in `test_utils`
module and are mainly used to facilitate testing.

ValuedMammal commented:
I cherry picked some of the changes from #1644. The main difference here is removing `insert_tx` from the wallet module and replacing it with similar functionality in `test_utils`. Despite the size of the PR, most of the changes are internal refactoring and adding documentation.

notmandatory reviewed (approved):
ACK b0dc3ddd3a5f2ff8e5b2fc5654fc2519bc7e728e

notmandatory commented:
This also fixes #1622 by removing functions that were only needed for testing.
